waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/AbstractThriftMetastoreClientManager.java

```diff
@@ -0,0 +1,174 @@
+      String[] metastoreUrisString = msUri.split(",");
+      metastoreUris = new URI[metastoreUrisString.length];
+      try {
+        int i = 0;
+        for (String s : metastoreUrisString) {
+          URI tmpUri = new URI(s);
+          if (tmpUri.getScheme() == null) {
+            throw new IllegalArgumentException("URI: " + s + " does not have a scheme");
+          }
+          metastoreUris[i++] = tmpUri;
+        }
+      } catch (IllegalArgumentException e) {
+        throw (e);
+      } catch (Exception e) {
+        String exInfo = "Got exception: " + e.getClass().getName() + " " + e.getMessage();
+        log.error(exInfo, e);
+        throw new RuntimeException(exInfo, e);
+      }
+    } else {
+      log.error("NOT getting uris from conf");
+      throw new RuntimeException("MetaStoreURIs not found in conf file");
+    }
+  }
+
+  void open() {
+    open(null);
+  }
+
+  abstract void open(HiveUgiArgs ugiArgs);
+
+  void reconnect(HiveUgiArgs ugiArgs) {
+    close();
+    // Swap the first element of the metastoreUris[] with a random element from the rest
+    // of the array. Rationale being that this method will generally be called when the default
+    // connection has died and the default connection is likely to be the first array element.
+    promoteRandomMetaStoreURI();
+    open(ugiArgs);
+  }
+
+  public String getHiveConfValue(String key, String defaultValue) {
+    return conf.get(key, defaultValue);
+  }
+
+  public void setHiveConfValue(String key, String value) {
+    conf.set(key, value);
+  }
+
+  public String generateNewTokenSignature(String defaultTokenSignature) {
+    String tokenSignature = conf.get(ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
+        defaultTokenSignature);
+    conf.set(ConfVars.METASTORE_TOKEN_SIGNATURE.varname,
+        tokenSignature);
+    return tokenSignature;
+  }
+
+  public Boolean isSaslEnabled() {
+    return conf.getBoolVar(ConfVars.METASTORE_USE_THRIFT_SASL);
+  }
+
+  @Override
+  public void close() {
+    if (!isConnected) {
+      return;
+    }
+    isConnected = false;
+    try {
+      if (client != null) {
+        client.shutdown();
+      }
+    } catch (TException e) {
+      log.debug("Unable to shutdown metastore client. Will try closing transport directly.", e);
+    }
+    // Transport would have got closed via client.shutdown(), so we don't need this, but
+    // just in case, we make this call.
+    if ((transport != null) && transport.isOpen()) {
+      transport.close();
+      transport = null;
+    }
+    log.info("Closed a connection to metastore, current connections: {}", CONN_COUNT.decrementAndGet());
+  }
+
+  boolean isOpen() {
+    return (transport != null) && transport.isOpen();
+  }
+
+  protected ThriftHiveMetastore.Iface getClient() {
+    return client;
+  }
+
+  /**
+   * Swaps the first element of the metastoreUris array with a random element from the remainder of the array.
+   */
+  private void promoteRandomMetaStoreURI() {
```

Comment:
So this is swapping the first metastore with the random metastore?

Reply:
These things are mostly how it was before it was an abstract class: https://github.com/ExpediaGroup/waggle-dance/pull/313/files#diff-7d969aea494084a6e71ce35f92df70de7b69337a9e08ef955d3ac2b5c4dc9170
Let's leave it, will me easier to merge master branch in at some point.